### PR TITLE
Models - Base, Account, AccountCategory, TaxType

### PR DESCRIPTION
**Pull request by darrynten**
## Base

The API calls are mostly _very_ generic, so there is a base model in place that
all other models extend off, which covers the following functionalities:

- GET Model/Get
- GET Model/Get/{id}
- DELETE Model/Delete/{id}
- POST Model/Save ** Incomplete **

This means that it's trivial to add new models that only use these calls (or a
combination of any of them) as there is a very simple 'recipe' to constructing
a basic model.

As such we only need to focus on the tricky bits.

### Basic model template

[Account Example Docs](https://accounting.sageone.co.za/api/1.1.2/Help/ResourceModel?modelName=Account)

There's a table at that link that looks something like this

```
Name             | Type             | Additional Information
-------------------------------------------------------------------
Name             | string           | None.
Category         | AccountCategory  | None.
Balance          | decimal          | Read Only / System Generated
ReportingGroupId | nullable integer | None.
```

We'll be using that for this example (docblocks excluded from example but are
required)

```php
/**
 * The name of the class is the `modelName=` value from the URL
 */
class Account extends BaseModel
{
    /**
     * Properties from the table, public
     *
     * We lowercase the first letter but leave the rest
     */
    public $id;
    public $name;
    public $reportingGroupId;

    // If something is read only we mark it in the docblock:
    /**
     * Balance
     *
     * READ ONLY
     *
     * @var double $balance
     */
    public $balance;

    // rest of properties...

    // The name of the endpoint (same as filename), protected
    protected $endpoint = 'Account';

    /**
     * Field definitions
     * 
     * Used by the base class to decide what gets submitted in a save call,
     * validation, etc
     *
     * All must include a type, whether or not it's nullable, and whether or
     * not it's persistable.
     *
     * - Nullable is `true` if the word 'nullable' is in the 'type' column
     * - Persistable is `true` if the word 'None.' is in the Additional Info column.
     * - Type has the following rules
     *   - `date` becomes "DateTime"
     *   - `nullable` is removed, i.e. "nullable integer" is only "integer"
     *   - Multiword linked terms are concatenated, eg:
     *     - "Account Category" becomes "AccountCategory"
     *     - "Tax Type" becomes "TaxType"
     *
     * Details on writable properties for Account:
     * https://accounting.sageone.co.za/api/1.1.2/Help/ResourceModel?modelName=Account
     * @var array $fields
     */
    protected $fields = [
        'id' => [
            'type' => 'integer',
            'nullable' => false,
            'persistable' => true,
        ],
        'name' => [
            'type' => 'string',
            'nullable' => false,
            'persistable' => true,
        ],
        'category' => [
            'type' => 'AccountCategory',
            'nullable' => false,
            'persistable' => true,
        ],
        'reportingGroupId' => [
            'type' => 'integer',
            'nullable' => true,
            'persistable' => true,
        ],
        'isTaxLocked' => [
            'type' => 'boolean',
            'nullable' => false,
            'persistable' => false,
        ],
        // etc etc etc
    ];

    /**
     * Features supported by the endpoint
     *
     * These features enable and disable the CRUD calls based on what is
     * supported by the SageOne API
     *
     * Most models use at least one of these features, with a fair amount
     * using all the functionality.
     *
     * @var array $features
     */
    protected $features = [
        'all' => true,
        'get' => true,
        'save' => true,
        'delete' => true,
    ];

    // Construct
    public function __construct(array $config)
    {
        parent::__construct($config);
    }

    // Add methods for non-standard calls, like:
    // GET Account/GetAccountsByCategoryId/{id}
    public function getAccountsByCategoryId($id)
    {
        // etc ...
    }
}
```

Following that template will very quickly create models for the project.

## Basic ORM-style mapping

Related models are auto-loaded and are all queryable, mutable, and persistable.

I'm sure there will be a recursion issue because of this at some point!

Some examples

```php
$account = new Account($config);

// get
$account->all(); // fetches ALL 
$account->get($id); // fetches that ID

// related models
echo $account->category->id;

// dates
echo $account->defaultTaxType->modified->format('Y-m-d');

// assign
$account->name = 'New Name';

// save, delete
$account->save(); // incomplete
$account->category->save(); // saving a child does not save the parent and vice-versa
$account->delete();
```

### CODE REVIEW

File: src/BaseModel.php [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    public function toJson()
+    {
+        return json_encode($this->toObject());
+    }
+
+    private function toObject()
+    {
+        $result = [];
```
Comment (darrynten) [inline src/BaseModel.php]: Please ignore this mess, it's being refactored, promise! :rofl:

File: src/BaseModel.php [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+                    $result[$remoteKey] = $this->$key;
+                } else {
+                    $class = $this->getModelWithNamespace($config['type']);
+
+                    if ($config['type'] === 'DateTime') {
+                        $result[$remoteKey] = $this->$key->format('Y-m-d');
+                    } elseif (!class_exists($class)) {
+                        die('class not exist');
+                    } else {
+                        // if (!$this->$key) {
```
Comment (mxnr) [inline src/BaseModel.php]: Missed comment

File: src/BaseModel.php [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                        }
+                        $result[$remoteKey] = $this->$key->toObject();
+                    }
+                }
+            }
+        }
+        return $result;
+    }
+
+    private function processResultItem($resultItem, $config)
```
Comment (mxnr) [inline src/BaseModel.php]: Phpdoc missed

File: src/BaseModel.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+        // Make a new instance of the class and load the item
+        $instance = new $class($this->config);
+
+        $instance->loadResult($resultItem);
+
+        return $instance;
+    }
+
+    public function loadResult(\stdClass $result)
```
Comment (mxnr) [inline src/BaseModel.php]: Phpdoc is missed

File: src/BaseModel.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+     *
+     * All methods that retrieve a single entity follow the same pattern, and this
+     * method here allows any compatible method to be called with something like:
+     *
+     * $account = new Account;
+     * $account->delete(11);
+     *
+     * @return void
+     */
+    public function delete(string $id)
```
Comment (mxnr) [inline src/BaseModel.php]: Param in doc is missed

File: src/BaseModel.php [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            // Load up the value from the object
+            $resultItem = $result->$remoteKey;
+
+            $value = $this->processResultItem($resultItem, $config);
+
+            $this->$key = $value;
+        }
+    }
+
+    private function getModelWithNamespace(string $model)
```
Comment (mxnr) [inline src/BaseModel.php]: And one more time:)

File: src/BaseModel.php [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        return sprintf(
```
Comment (mxnr) [inline src/BaseModel.php]: I think it's better to use `ClassNAME::class` in this case

File: src/Models/Account.php [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+     * READ ONLY
+     *
+     * @var boolean $isTaxLocked
+     */
+    public $isTaxLocked;
+
+    /**
+     * Date the account was last modified
+     *
+     * READ ONLY
```
Comment (mxnr) [inline src/Models/Account.php]: I think in this case you need to use `const`

Comment (mxnr) [inline src/Models/Account.php]: *i mean read only case

File: src/Models/AccountCategory.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    public $id;
+
+    /**
+     * The API Endpoint
+     *
+     * @var string $endpoint
+     */
+    protected $endpoint = 'AccountCategory';
+
+    public $comment;
```
Comment (mxnr) [inline src/Models/AccountCategory.php]: I would like to have comments here also

File: src/Request/RequestHandler.php [hunk 10]
```diff
@@ -110,7 +110,11 @@ public function __construct(array $config)
         $this->key = $config['key'];
         $this->endpoint = $config['endpoint'];
         $this->version = $config['version'];
-        $this->companyId = $config['companyId'] || null;
+
+        if (isset($config['companyId'])) {
```
Comment (mxnr) [inline src/Request/RequestHandler.php]: Array_key_exists should be used instead

File: src/Request/RequestHandler.php [hunk 11]
```diff
@@ -153,7 +157,7 @@ public function handleRequest(string $method, string $uri, array $options, array
             $this->handleException($exception);
         }
 
-        return json_decode($response->getBody(), true);
+        return json_decode($response->getBody(), false);
```
Comment (mxnr) [inline src/Request/RequestHandler.php]: Second param is false by default

File: src/BaseModel.php [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                }
+            }
+        }
+        return $result;
+    }
+
+    private function processResultItem($resultItem, $config)
+    {
+        // The type of the item matches the required type
+        if ($config['type'] === gettype($resultItem)) {
```
Comment (mxnr) [inline src/BaseModel.php]: What if this array wouldn't have this `type` key?

File: src/BaseModel.php [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+            return $resultItem;
+        }
+
+        // If it's a date we return a new DateTime object
+        if ($config['type'] === 'DateTime') {
```
Comment (mxnr) [inline src/BaseModel.php]: I think ClassName::class would be a better solution

File: src/BaseModel.php [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+            return new \DateTime($resultItem);
+        }
+
+        // At this stage, any type is going to be a model that needs to be loaded
+        $class = $this->getModelWithNamespace($config['type']);
+
+        // So if the class doesn't exist, throw
+        if (!class_exists($class)) {
+            die(var_dump('NO CLASS', $class, $config, gettype($resultItem), $resultItem));
+            throw new ModelException('Model contains a model property defined without a corresponding class', 110);
```
Comment (mxnr) [inline src/BaseModel.php]: This condition never be reached

File: src/Models/Account.php (hunk 8 above)
Comment (darrynten) [inline src/Models/Account.php]: These aren't read only in that way, they're read only on the Sage API side, i.e. can't persist directly through that model.

_Sent from my LGE Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

File: src/BaseModel.php (hunk 12 above)
Comment (darrynten) [inline src/BaseModel.php]: Having the type key is a hard required, but perhaps a better way?

_Sent from my LGE Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

File: src/BaseModel.php (hunk 14 above)
Comment (darrynten) [inline src/BaseModel.php]: Refactoring this out now 😎

_Sent from my LGE Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by darrynten** [comment]
NB Save is not implemented yet

**Comment by darrynten** [comment]
*phew* Ok, ready for re-review @mxnr 🦄

File: src/BaseModel.php (hunk 13 above)
Comment (darrynten) [inline src/BaseModel.php]: Implemented :)

**Pull request merged**